File: python/system_design/message_queue.py
from collections import deque
class Topic: __init__(self, name): self.name=name; self.queue=deque(); self.subs=[]
class MessageQueue:
    def __init__(self): self.topics={}
    def create_topic(self, t): self.topics[t]=Topic(t)
    def publish(self, t, m): self.topics[t].queue.append(m)
    def subscribe(self, t, f): self.topics[t].subs.append(f)
    def consume(self, t): return list(self.topics[t].queue) if t in self.topics else []
if __name__ == "__main__": mq=MessageQueue(); mq.create_topic("t1"); mq.publish("t1","msg1"); print(mq.consume("t1"))

File: java/system_design/MessageQueue.java
import java.util.*;
/**
 * MessageQueue - [Brief description]
 *
 * <p>OVERVIEW:
 * [Detailed explanation of what this class does]
 *
 * <p>COMPLEXITY:
 * <ul>
 *   <li>Time: [See method documentation]</li>
 *   <li>Space: O(n) where n is [the element count]</li>
 * </ul>
 *
 * <p>USAGE:
 * [How to use this class, with example]
 *
 * @author Interview Preparation
 * @since 1.0
 */

/**
 * [Brief description]
 *
 * @param [param] [description]
 * @return [description]
 * @time O([complexity])
 */
public class MessageQueue{class Topic{Queue<String>q=new LinkedList<>();}Map<String,Topic>topics=new HashMap<>();void createTopic(String t){topics.put(t,new Topic());}void publish(String t,String m){topics.get(t).q.add(m);}public static void main(String[]a){MessageQueue mq=new MessageQueue();mq.createTopic("t1");mq.publish("t1","msg");}}
